internal/receiver/smartagentreceiver/log.go

```diff
@@ -0,0 +1,167 @@
+		// No zap level equivalent to trace. Mapping trace to debug.
+		logrus.TraceLevel: zapcore.DebugLevel,
+	}
+)
+
+func (l *logrusKey) reportCaller() {
+	if !l.ReportCaller {
+		l.SetReportCaller(true)
+	}
+}
+
+func (l *logrusKey) discardOutput() {
+	if l.Out != ioutil.Discard {
+		l.SetOutput(ioutil.Discard)
+	}
+}
+
+func (l *logrusKey) hookUnique(newHook logrus.Hook) {
+	for _, hooks := range l.Hooks {
+		for _, hook := range hooks {
+			if hook == newHook {
+				return
+			}
+		}
+	}
+	l.AddHook(newHook)
+}
+
+func (l *logrusToZap) redirect(src logrusKey, dst *zap.Logger) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if _, ok := l.loggerMap[src]; !ok {
+		l.loggerMap[src] = make([]*zap.Logger, 0)
+	}
+
+	// mutating the logrus logger as a side effect.
+	src.reportCaller()
+	src.discardOutput()
+	src.hookUnique(l)
+
+	l.loggerMap[src] = append(l.loggerMap[src], dst)
+}
+
+func (l *logrusToZap) unRedirect(src logrusKey, dst *zap.Logger) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if _, ok := l.loggerMap[src]; !ok {
+		return
+	}
+
+	keep := make([]*zap.Logger, 0)
+
+	for _, logger := range l.loggerMap[src] {
+		if logger != dst {
+			keep = append(keep, logger)
+		}
+	}
+
+	l.loggerMap[src] = keep
+}
+
+func (l *logrusToZap) loggerMapValue0(src logrusKey) *zap.Logger {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if loggers, ok := l.loggerMap[src]; !ok || len(loggers) == 0 {
+		return nil
+	}
+
+	return l.loggerMap[src][0]
+}
+
+// Levels is a logrus.Hook implementation that returns all logrus logging levels.
+func (l *logrusToZap) Levels() []logrus.Level {
+	return logrus.AllLevels
+}
+
+// Fire is a logrus.Hook implementation that is called when logging on the logging levels returned by Levels.
+// A zap log entry is created from the supplied logrus entry and written out.
+func (l *logrusToZap) Fire(e *logrus.Entry) error {
+	var monitorType string
+
+	fields := make([]zapcore.Field, 0)
+
+	// Creating zap entry fields from logrus entry fields.
+	for k, v := range e.Data {
+		vStr := fmt.Sprintf("%v", v)
+		if k == "monitorType" {
+			monitorType = vStr
+		}
+		fields = append(fields, zapcore.Field{Key: k, Type: zapcore.StringType, String: vStr})
+	}
+
+	// Creating loggerMap key from the logrus entry logger and field 'monitorType'
+	key := logrusKey{e.Logger, monitorType}
+	zapLogger := l.loggerMapValue0(key)
+	if zapLogger == nil {
+		fields = append(fields, zapcore.Field{Key: "monitorType", Type: zapcore.StringType, String: monitorType})
+		fields = append(fields, zapcore.Field{Key: "redirect_error", Type: zapcore.StringType, String: "Could not find zap logger in receiver for the monitorType. Using the catchall zap logger instead."})
+		zapLogger = l.catchallLogger
+	}
+
+	zapLevel := levelsMap[e.Level]
+	ce := zapLogger.Check(zapLevel, e.Message)
+	if ce == nil {
+		ce = zapLogger.Check(zap.ErrorLevel, fmt.Sprintf("Failed to redirect logrus logs at level %s. The matching zap log level %s is not enabled.", e.Level.String(), zapLevel.String()))
```

Comment:
This logs error statements for every debug log statement if debug logging isn't enabled?

Reply:
If debug logging is enabled for logrus and disabled for zap then yes.

Reply:
Right, couldn't that flood logs with undesired errors?  Why not just not log the statements?